File: src/network/models/grpc_response.py
from __future__ import annotations
from dataclasses import dataclass
from google.protobuf.descriptor import FileDescriptor

from network.models.flow_response import FlowResponse
from network.utils_grpc import decode_grpc_data, extract_grpc_path_info


@dataclass(kw_only=True)
class GrpcResponse(FlowResponse):
    headers: dict[str, list[str]]
    body: bytes

    def __str__(self):
        out = ""
        for key, value in self.headers.items():
            v = ",".join(value)
            out += f"{key}: {v}\n"

        out += "\r\n"
        if len(self.body) > 0:
            out += self.body.decode()

        return out

    def decode_body(self, file_descriptor: FileDescriptor, path: str) -> str:
        _, service_name, method_name = extract_grpc_path_info(path)

        service = file_descriptor.services_by_name[service_name]
        method = service.methods_by_name[method_name]
        msg_name = method.output_type.name

        return decode_grpc_data(self.body, file_descriptor, msg_name)

File: lib/network/models/grpc_response.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:trayce/common/utils.dart';
import 'package:trayce/network/models/proto_def.dart';

import '../../agent/gen/api.pb.dart' as pb;
import 'flow_response.dart';

/// gRPC response with parsed fields
class GrpcResponse extends FlowResponse {
  final Map<String, List<String>> headers;
  final Uint8List body;

  const GrpcResponse({
    required this.headers,
    required this.body,
  }) : super();

  /// Creates a GrpcResponse from raw JSON bytes
  static GrpcResponse fromJson(Uint8List bytes) {
    final values = json.decode(utf8.decode(bytes)) as Map<String, dynamic>;
    return GrpcResponse(
      headers: (values['headers'] as Map<String, dynamic>).map(
        (key, value) => MapEntry(key, (value as List).cast<String>()),
      ),
      body: Uint8List.fromList(base64.decode(values['payload'] as String)),
    );
  }

  /// Creates a GrpcResponse from a protobuf GRPCResponse
  factory GrpcResponse.fromProto(pb.GRPCResponse res) {
    // Convert headers from protobuf repeated fields to Map<String, List<String>>
    final headers = <String, List<String>>{};
    for (var entry in res.headers.entries) {
      headers[entry.key] = entry.value.values.map((v) => v.toString()).toList();
    }

    return GrpcResponse(
      headers: headers,
      body: Uint8List.fromList(res.payload),
    );
  }

  @override
  Uint8List toJson() {
    final map = {
      'headers': headers,
      'payload': base64.encode(body),
    };
    return utf8.encode(json.encode(map));
  }

  @override
  String toString() {
    var out = formatSortedHeaders(headers);

    out += '\r\n';
    if (body.isNotEmpty) {
      out += utf8.decode(body);
    }

    return out;
  }

  String toStringParsed(ProtoDef protoDef, String grpcMsgPath) {
    var out = formatSortedHeaders(headers);

    out += '\r\n';
    if (body.isNotEmpty) {
      try {
        final parsedBody = protoDef.parseGRPCMessage(body, grpcMsgPath, true);
        // Try to parse the body as JSON and format it with indentation
        final jsonObj = json.decode(parsedBody);
        final prettyJson = const JsonEncoder.withIndent('  ').convert(jsonObj);
        out += '\n$prettyJson';
      } catch (e) {
        print(e);
        out += '\nFailed to parse using ${protoDef.name}';
      }
    }

    return out;
  }

  @override
  String responseCol() {
    return "ok";
  }
}
